### DISCUSSION TITLE
How to enable full linker on net6.0-android and net6.0-ios so that Refit and System.Text.Json work?

### DISCUSSION THREAD

**Discussion by jeanplevesque** [unoplatform/uno#11214] [Q&A]
We're building a mobile net6.0 app using Uno.
It uses System.Text.Json (with source generation) and Refit (which also uses source generation).
When building with `MtouchLink`/`AndroidLinkMode` to `Full`, the app runs and gets runtime exceptions.
![image](https://user-images.githubusercontent.com/39710855/216990018-7177c2c2-7361-434a-8323-9c190d92570c.png)
![image](https://user-images.githubusercontent.com/39710855/216990101-6190505a-ec31-439b-ab2f-2ffcc7f9bc44.png)
When building with `None` or `SdkOnly`, the app runs without exceptions.

The assembly that has this code generation (ApplicationTemplate.Client) is listed in LinkerExclusions.xml so I don't understand why enabling and disabling the linker has an impact.

![image](https://user-images.githubusercontent.com/39710855/216991637-ddf05a22-21a8-4874-af81-5db7388f116c.png)

To repro, you can build and run this application in debug (on either iOS or Android). Make sure you enable all exceptions to see them.
[UnoApplicationTemplate.zip](https://github.com/unoplatform/uno/files/10664114/UnoApplicationTemplate.zip)

**Comment by jeromelaban** [comment]
See [this Microsoft](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0#root-descriptors) article on trimming configuration for the linker. You looking for `TrimmerRootDescriptor`.

**Reply by jeanplevesque** [reply]
What's the difference between the linker and trimmer? Are they the same thing?

**Reply by jeromelaban** [reply]
They're the same tool, just that the trimmer is bundled with the .NET SDK.

**Reply by jeanplevesque** [reply]
Ok. Right now we use the _Xamarin-named_ properties (`MtouchLink`/`AndroidLinkMode`) and they have an impact. Are we supposed to stop using those and replace them with the new properties (`PublishTrimmed`, `TrimMode`, etc.)?

**Reply by jeromelaban** [reply]
They are both active at this time, but in the future the Xamarin properties will likely be removed.

**Reply by jeanplevesque** [reply]
> They are both active at this time

What does that mean?
Do we need to use both set of options? Should we only use the new options? Only the old ones? Are they conflicting? Are they equivalent?

How are the Uno .Net 6 mobile apps optimized?

**Reply by jeromelaban** [reply]
Both options point to the same tooling either ones can be currently used, with the same effects.

If you still have issues with the new parameters, can you explain what these are?

**Reply by jeanplevesque** [reply]
1. When I don't enable trimming, the app works fine (and the APK is larger).
2. When I enable trimming, deserialization doesn't work (and the APK is smaller).
   - The exception is `**System.InvalidOperationException:** 'ConstructorParamIncompleteBinding, ApplicationTemplate.Client.AuthenticationToken'`.
   - We use `[JsonSerializable(typeof(XYZ))]` to leverage the code-generated serializers of System.Text.Json.
   - I tried 3 ways of enabling the full linker. (Deserialization fails in all 3.)
     1. ```xml
        <AndroidLinkMode>Full</AndroidLinkMode>
        <LinkDescription Include="Android\LinkerExclusions.xml" />
        ```
     1. ```xml
        <PublishTrimmed>true</PublishTrimmed>
        <TrimMode>link</TrimMode>
        <EmbeddedResource Include="Android\LinkerExclusions.xml"
           <LogicalName>ILLink.Descriptors.xml</LogicalName>
        </EmbeddedResource>
        ```
     1. ```xml
        <PublishTrimmed>true</PublishTrimmed>
        <TrimMode>link</TrimMode>
        <TrimmerRootDescriptor Include="Android\LinkerExclusions.xml" />
        ```
   - With configurations `a.` and `b.`, we also get an exception from Refit.
     ```
     [Uno.UI.DataBinding.BindingPath+BindingItem] Failed to get the source value for [Jokes] System.Reflection.TargetInvocationException: Arg_TargetInvocationException
     [Uno.UI.DataBinding.BindingPath+BindingItem]  ---> System.ArgumentException: URL /{typePost}.json has parameter typepost, but no method parameter matches
     [Uno.UI.DataBinding.BindingPath+BindingItem]    at Refit.RestMethodInfo.BuildParameterMap(String , List`1 ) in d:\a\1\s\Refit\RestMethodInfo.cs:line 191
     [Uno.UI.DataBinding.BindingPath+BindingItem]    at Refit.RestMethodInfo..ctor(Type , MethodInfo , RefitSettings ) in d:\a\1\s\Refit\RestMethodInfo.cs:line 72
     [Uno.UI.DataBinding.BindingPath+BindingItem]    at Refit.RequestBuilderImplementation.AddInterfaceHttpMethods(Type , Dictionary`2 ) in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 77
     [Uno.UI.DataBinding.BindingPath+BindingItem]    at Refit.RequestBuilderImplementation..ctor(Type , RefitSettings ) in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 55
     [Uno.UI.DataBinding.BindingPath+BindingItem]    at Refit.RequestBuilderImplementation`1[[ApplicationTemplate.Client.IDadJokesEndpoint, ApplicationTemplate.Client, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]..ctor(RefitSettings ) in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 20
     [Uno.UI.DataBinding.BindingPath+BindingItem]    at Refit.RequestBuilderFactory.Create[IDadJokesEndpoint](RefitSettings ) in d:\a\1\s\Refit\RequestBuilderFactory.cs:line 18
     [Uno.UI.DataBinding.BindingPath+BindingItem]    at Refit.RequestBuilder.ForType[IDadJokesEndpoint](RefitSettings ) in d:\a\1\s\Refit\RequestBuilder.cs:line 24
     ...
     ```
   - LinkerExclusions.xml contains the following.
     ```xml
     <?xml version="1.0" encoding="utf-8" ?>
     <linker>
        <assembly fullname="System">
           <type fullname="System.Diagnostics.Debug">
              <method name="WriteLine" />
           </type>
        </assembly>

        <assembly fullname="Serilog.Extensions.Hosting" />
	
        <assembly fullname="Microsoft.Extensions.Hosting" />
        <assembly fullname="Microsoft.Extensions.Options" />
        <assembly fullname="Microsoft.Extensions.Http" />

        <assembly fullname="ApplicationTemplate.Client" />
        <assembly fullname="ApplicationTemplate.Business" />
        <assembly fullname="ApplicationTemplate.Presentation" />

        <assembly fullname="System.Text.Json" />
     </linker>
     ```

**Reply by jeanplevesque** [reply]
I created an documentation issue: https://github.com/unoplatform/uno/issues/11248

**Reply by jeromelaban** [reply]
See a workaround for this specific issue in https://github.com/unoplatform/uno/issues/11248#issuecomment-1426399900.

**Comment by teo-tsirpanis** [comment]
Linker-friendliness of Refit is tracked in reactiveui/refit#1389.
